File: SSUMobile/Modules/News/Views/SSUNewsViewController.m
//
//  SSUNewsViewController.m
//  SSUMobile
//
//  Created by Andrew Huss on 2/1/13.
//  Copyright (c) 2013 Sonoma State University Department of Computer Science. All rights reserved.
//

@import SafariServices;

#import "SSUNewsViewController.h"
#import "SSUSelectionController.h"
#import "SSULogging.h"
#import "SSUNewsModule.h"
#import "SSUNewsConstants.h"
#import "SSUArticle.h"
#import "SSUNewsArticleTableViewCell.h"
#import "SSUWebViewController.h"
#import "SSUConfiguration.h"

@interface SSUNewsViewController () <SSUSelectionDelegate>

@property (nonatomic, strong) NSManagedObjectContext * context;
@property (nonatomic, strong) NSPredicate * predicate;
@property (nonatomic, strong) UIBarButtonItem * filterButton;

@end

@implementation SSUNewsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.context = [[SSUNewsModule sharedInstance] context];
    [self.refreshControl addTarget:self action:@selector(refresh) forControlEvents:UIControlEventValueChanged];
    
    self.filterButton = [[UIBarButtonItem alloc] initWithTitle:@"Filter" style:UIBarButtonItemStyleDone target:self action:@selector(filterButtonPressed)];
    UIBarButtonItem * flexibleWidthButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:NULL];
    self.toolbarItems = @[flexibleWidthButton, self.filterButton, flexibleWidthButton];
    
    self.fetchedResultsController = [self makeFetchedResultsController];
    self.searchFetchedResultsController = [self makeFetchedResultsController];
    self.searchKey = @"title";
    
    [self.tableView registerClass:[SSUNewsArticleTableViewCell class]
           forCellReuseIdentifier:SSUNewsEntityArticle];
    self.tableView.separatorInset = UIEdgeInsetsZero;
    self.tableView.rowHeight = (int)([UIScreen mainScreen].applicationFrame.size.height / 4);
    [self.tableView reloadData];
}

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    const NSTimeInterval lastUpdate = [[[SSUConfiguration sharedInstance] dateForKey:SSUNewsUpdatedDateKey] timeIntervalSinceNow];
    const NSTimeInterval timeInterval = -1 * 60 * 5;
    if (lastUpdate <= timeInterval) {
        [self refresh];
    }
    self.navigationController.toolbarHidden = NO;
}

- (void) viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    self.navigationController.toolbarHidden = YES;
}

- (void)refresh {
    SSULogDebug(@"Refresh News");
    if (!self.refreshControl.refreshing)
        [self.refreshControl beginRefreshing];
    [[SSUNewsModule sharedInstance] updateData:^{
        [self.refreshControl endRefreshing];
        [self.tableView reloadData];
    }];
}

- (void) setPredicate:(NSPredicate *)predicate {
    _predicate = predicate;
    self.fetchedResultsController.fetchRequest.predicate = _predicate;
    [self.fetchedResultsController performFetch:nil];
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = nil;
    if (!cellIdentifier) cellIdentifier = SSUNewsEntityArticle;
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier forIndexPath:indexPath];
    [self configureCell:cell atIndexPath:indexPath];
    return cell;
}

- (void)configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath {
    SSUArticle *article = (SSUArticle *)[self objectAtIndex:indexPath];
    if ([cell.reuseIdentifier isEqualToString:SSUNewsEntityArticle]) {
        SSUNewsArticleTableViewCell* articleCell = (SSUNewsArticleTableViewCell*)cell;
        articleCell.article = article;
        articleCell.separatorInset = UIEdgeInsetsZero;
        articleCell.layoutMargins = UIEdgeInsetsZero;
    }
}

- (NSString*) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return nil;
}

- (UIView *) tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    return nil;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    SSUArticle* article = [self.currentFetchedResultsController objectAtIndexPath:indexPath];
    NSURL * url = [NSURL URLWithString:article.link];
    SSULogDebug(@"%@",article);
    if (NSStringFromClass([SFSafariViewController class])) {
        SFSafariViewController * vc = [[SFSafariViewController alloc] initWithURL:url entersReaderIfAvailable:YES];
        vc.preferredBarTintColor = SSU_BLUE_COLOR;
        [self presentViewController:vc animated:YES completion:NULL];
    }
    else {
        SSUWebViewController* controller = [SSUWebViewController webViewController];
        controller.urlToLoad = [NSURL URLWithString:article.link];
        [self.navigationController pushViewController:controller animated:YES];
    }
}

- (NSPredicate *) searchPredicateForText:(NSString *)searchText {
    NSPredicate * base = [self basePredicate];
    NSPredicate * search = [super searchPredicateForText:searchText];
    if (self.predicate != nil) {
        return [NSCompoundPredicate andPredicateWithSubpredicates:@[base, search, self.predicate]];
    } else {
        return [NSCompoundPredicate andPredicateWithSubpredicates:@[base, search]];
    }
}

#pragma mark - Actions

- (void) filterButtonPressed {
    NSArray * categories = [@[@"All Categories"] arrayByAddingObjectsFromArray:[self allCategories]];
    SSUSelectionController * selectionController = [[SSUSelectionController alloc] initWithItems:categories];
    selectionController.delegate = self;
    if (self.predicate == nil) {
        selectionController.defaultIndex = [NSIndexPath indexPathForRow:0 inSection:0];
    }
    else {
        NSString * current = [[self.fetchedResultsController objectAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]] category];
        NSInteger row = (NSInteger)[categories indexOfObject:current];
        selectionController.defaultIndex = [NSIndexPath indexPathForRow:row inSection:0];
    }
    [self.navigationController pushViewController:selectionController animated:YES];
}

- (void) userDidSelectItem:(id)item atIndexPath:(NSIndexPath *)indexPath fromController:(SSUSelectionController *)controller {
    if ([item isEqualToString:@"All Categories"]) {
        self.predicate = nil;
        self.filterButton.title = @"Filter";
    }
    else {
        NSString * categoryName = item;
        self.filterButton.title = categoryName;
        self.predicate = [NSPredicate predicateWithFormat:@"category = %@", categoryName];
    }
    
    [self.navigationController popViewControllerAnimated:YES];
}

- (void) selectionControllerDismissed:(SSUSelectionController *)controller {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - private

- (NSPredicate *) basePredicate {
    // Don't fetch articles older than 6 months
    NSDate * oldestToFetch = [[NSDate date] dateByAddingTimeInterval:-1*ABS(SSUNewsArticleFetchDateLimit)];
    return [NSPredicate predicateWithFormat:@"published > %@",oldestToFetch];
}

- (NSFetchedResultsController *) makeFetchedResultsController {
    NSFetchRequest* fetchRequest = [[NSFetchRequest alloc] init];
    fetchRequest.entity = [NSEntityDescription entityForName:SSUNewsEntityArticle inManagedObjectContext:self.context];
    fetchRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"published" ascending:NO]];
    fetchRequest.predicate = [self basePredicate];
    
    NSFetchedResultsController *controller = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest
                                                                                 managedObjectContext:self.context
                                                                                   sectionNameKeyPath:@"published"
                                                                                            cacheName:nil];
    return controller;
}

- (NSArray *) allCategories {
    NSFetchRequest * request = [NSFetchRequest fetchRequestWithEntityName:SSUNewsEntityArticle];
    NSString * categoryKeyName = @"category";
    request.propertiesToFetch = @[categoryKeyName];
    request.returnsDistinctResults = YES;
    request.resultType = NSDictionaryResultType;
    
    NSArray * objects = [[SSUNewsModule sharedInstance].context executeFetchRequest:request error:nil];
    NSMutableArray * categories = [NSMutableArray arrayWithCapacity:objects.count];
    for (NSDictionary * dict in objects) {
        if (dict[categoryKeyName] != nil) {
            [categories addObject:dict[categoryKeyName]];
        }
    }
    return [categories sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
}

@end

File: SSUMobile/Modules/News/Views/SSUNewsViewController.swift
//
//  SSUNewsViewController.swift
//  SSUMobile
//
//  Created by Eric Amorde on 3/26/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation
import SafariServices

class SSUNewsViewController: SSUCoreDataTableViewController, SSUSelectionDelegate {
    
    private let cellIdentifier: String = "SSUArticle"
    
    private let autoUpdateDelay: TimeInterval = 60 * 5
    
    var predicate: NSPredicate? {
        didSet {
            fetchedResultsController?.fetchRequest.predicate = predicate
            performFetch()
            tableView.reloadData()
        }
    }
    
    var filterButton: UIBarButtonItem!
    var context: NSManagedObjectContext!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "University News"
        refreshControl = UIRefreshControl()
        refreshControl?.addTarget(self, action: #selector(self.refresh), for: .valueChanged)
        
        context = SSUNewsModule.instance.context
        fetchedResultsController = self.makeFetchedResultsController()
        searchFetchedResultsController = self.makeFetchedResultsController()
        searchKey = "title"
        
        filterButton = UIBarButtonItem(title: "Filter", style: .done, target: self, action: #selector(self.filterButtonPressed))
        let flexibleButton = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
        toolbarItems = [flexibleButton, filterButton, flexibleButton]
        
        tableView.register(SSUNewsArticleTableViewCell.self, forCellReuseIdentifier: cellIdentifier)
        tableView.separatorInset = UIEdgeInsets.zero
        tableView.estimatedRowHeight = round(UIScreen.main.bounds.height / 4.0)
        tableView.reloadData()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.isToolbarHidden = false
        if let lastUpdate = SSUConfiguration.sharedInstance().newsLastUpdate {
            if abs(lastUpdate.timeIntervalSinceNow) > autoUpdateDelay {
                refresh()
            }
        } else {
            refresh()
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        navigationController?.isToolbarHidden = true
    }
    
    // MARK: UITableView
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as? SSUNewsArticleTableViewCell else {
            SSULogging.logError("Unable to dequeue news article cell")
            return UITableViewCell()
        }
        
        cell.article = self.object(atIndex: indexPath) as? SSUArticle
        cell.separatorInset = .zero
        cell.layoutMargins = .zero
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        guard let article = object(atIndex: indexPath) as? SSUArticle,
            let link = article.link,
            let articleURL = URL(string: link) else {
            SSULogging.logError("Unable to retrieve article of selected row or article is missing link")
            return
        }
        
        if #available(iOS 9.0, *) {
            let vc = SFSafariViewController(url: articleURL, entersReaderIfAvailable: true)
            if #available(iOS 10.0, *) {
                vc.preferredBarTintColor = .ssuBlue
            }
            present(vc, animated: true, completion: nil)
        } else {
            let vc = SSUWebViewController()
            vc.urlToLoad = articleURL
            navigationController?.pushViewController(vc, animated: true)
        }
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        return nil
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 1.0
    }
    
    // MARK: Actions
    
    @objc
    func refresh() {
        SSULogging.logDebug("Refresh News")
        if !(refreshControl?.isRefreshing ?? false) {
            refreshControl?.beginRefreshing()
        }
        SSUNewsModule.instance.updateData {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self.refreshControl?.endRefreshing()
            }
            self.tableView.reloadData()
        }
    }
    
    @objc
    func filterButtonPressed() {
        var categories: [String] = ["All Categories"]
        categories.append(contentsOf: allCategories())
        
        let selectionVc = SSUSelectionController(items: categories)!
        selectionVc.delegate = self
    }
    
    // MARK: SSUSelectionDelegate
    
    func userDidSelectItem(_ item: Any!, at indexPath: IndexPath!, from controller: SSUSelectionController!) {
        let category = item as! String
        if category == "All Categories" {
            predicate = nil
            filterButton.title = "Filter"
        } else {
            filterButton.title = category
            predicate = NSPredicate(format: "category = %@", category)
        }
        
        let _ = navigationController?.popViewController(animated: true)
    }
    
    func selectionControllerDismissed(_ controller: SSUSelectionController!) {
        let _ = navigationController?.popViewController(animated: true)
    }
    
    // MARK: Search
    
    override func searchPredicate(forText searchText: String) -> NSPredicate {
        let base = basePredicate()
        let search = super.searchPredicate(forText: searchText)
        if let existingPredicate = self.predicate {
            return NSCompoundPredicate(andPredicateWithSubpredicates: [base, search, existingPredicate])
        } else {
            return NSCompoundPredicate(andPredicateWithSubpredicates: [base, search])
        }
    }
    
    // MARK: NSFetchedResultsController
    
    private func basePredicate() -> NSPredicate {
        let fetchCutoff = Date(timeIntervalSinceNow: -1*abs(SSUNewsModule.instance.articleFetchDateLimit))
        return NSPredicate(format: "published >= %@", argumentArray: [fetchCutoff])
    }
    
    func makeFetchedResultsController() -> NSFetchedResultsController<NSFetchRequestResult> {
        let fetchRequest: NSFetchRequest<SSUArticle> = SSUArticle.fetchRequest()
        fetchRequest.sortDescriptors = [
            NSSortDescriptor(key: "published", ascending: false)
        ]
        fetchRequest.predicate = basePredicate()
        let controller = NSFetchedResultsController(fetchRequest: fetchRequest,
                                                    managedObjectContext: context,
                                                    sectionNameKeyPath: "published",
                                                    cacheName: nil)
        return controller as! NSFetchedResultsController<NSFetchRequestResult>
    }
    
    // MARK: Helper
    
    private func allCategories() -> [String] {
        let request = NSFetchRequest<NSDictionary>(entityName: "SSUArticle")
        let keyName = "category"
        request.propertiesToFetch = [keyName]
        request.returnsDistinctResults = true
        request.resultType = .dictionaryResultType
        
        if let results = try? context.fetch(request) {
            let categories = results.flatMap({ $0[keyName] as? String })
            return categories.sorted()
        }
        
        return []
    }
}
